File: lab2/KotlinArray/parser/Tree.cpp
#include <fstream>
#include "Tree.h"


Tree::Tree(Tree::Terminal terminal)
    : node(terminal)
{}

Tree::Tree(NonTerminal nonTerminal)
    : node(nonTerminal)
{}

Tree::~Tree() {
    clear();
}


void Tree::addChild(Tree* child) {
    children.push_back(child);
}

void Tree::clear() {
    for (auto ptr : children) {
        delete ptr;
    }
}

std::vector<Tree::Terminal> Tree::walkthrough() {
    std::vector<Terminal> result;
    walkthroughImpl(result, this);
    return result;
}

void Tree::walkthroughImpl(std::vector<Terminal>& result, Tree* tree) {
    if (tree->children.empty()) {
        result.push_back(std::get<Terminal>(tree->node));
        return;
    }
    for (auto child : tree->children) {
        walkthroughImpl(result, child);
    }
}

void Tree::walkthroughGraphviz(std::string path) {
    std::vector<std::pair<size_t, size_t>> edges;
    std::vector<std::string> nodeConfig;

    walkthroughGraphvizImpl(edges, nodeConfig, this);

    std::ofstream out(path);
    out << "strict graph KotlinArray {\n";
    for (std::size_t i = 0; i != nodeConfig.size(); ++i) {
        out << "    " << i + 1 << "[" << nodeConfig[i] << "];\n";
    }
    for (auto [lhs, rhs] : edges) {
        out << "    " << lhs + 1 << " -- " << rhs + 1 << ";\n";
    }
    out << "}\n";
    out.close();
}

void Tree::walkthroughGraphvizImpl(std::vector<std::pair<size_t, size_t>>& edges,
                                   std::vector<std::string>& nodeConfig,
                                   Tree* tree) {
    std::size_t treeID = nodeConfig.size();
    nodeConfig.emplace_back();

    if (tree->children.empty()) {
        nodeConfig[treeID] += "label=\"";
        nodeConfig[treeID] += Token2String[std::get<Terminal>(tree->node)];
        nodeConfig[treeID] += "\" shape=box color=green";
        return;
    } else {
        nodeConfig[treeID] += "label=\"";
        nodeConfig[treeID] += NonTerminal2String[std::get<NonTerminal>(tree->node)];
        nodeConfig[treeID] += "\" shape=circle";
    }

    for (auto child : tree->children) {
        std::size_t childID = nodeConfig.size();
        edges.emplace_back(treeID, childID);
        walkthroughGraphvizImpl(edges, nodeConfig, child);
    }
}

File: lab4/TranslatorGenerator/src/main/java/parser/Tree.java
package parser;

import grammar.GrammarItem;
import grammar.NonTerminal;
import grammar.Terminal;
import grammar.Type;

import java.io.BufferedWriter;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStreamWriter;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Tree {
    private final GrammarItem node;
    private final List<Tree> children = new ArrayList<>();
    private Type type;

    public Map<String, String> synthesizedAttr = new HashMap<>();


    private class Edge {
        private final int from;
        private final int to;

        public Edge(final int from, final int to) {
            this.from = from;
            this.to = to;
        }

        public int getFrom() {
            return from;
        }

        public int getTo() {
            return to;
        }
    }


    public Tree(final GrammarItem node) {
        this.node = node;
    }

    public Tree(final GrammarItem node, final Type type) {
        this.node = node;
        this.type = type;
    }

    public GrammarItem getNode() {
        return node;
    }

    public void addChild(final Tree child) {
        children.add(child);
    }

    public void setType(Type type) {
        this.type = type;
    }

    public void walkthroughGraphviz(String path) {
        final List<Edge> edges = new ArrayList<>();
        final List<String> nodeConfig = new ArrayList<>();

        walkthroughGraphvizImpl(edges, nodeConfig, this);

        try (BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(new FileOutputStream(path), StandardCharsets.UTF_8))) {
            writer.write("strict graph TranslatorGenerator {\n");
            for (int i = 0; i != nodeConfig.size(); ++i) {
                writer.write(String.format("    %d[%s];\n", i + 1, nodeConfig.get(i)));
            }
            for (Edge edge : edges) {
                writer.write(String.format("    %d -- %d;\n", edge.getFrom() + 1, edge.getTo() + 1));
            }
            writer.write("}\n");
        } catch (IOException e) {
            System.err.printf("Error occurred while writing result into file '%s'", path);
        }
    }


    private void walkthroughGraphvizImpl(List<Edge> edges, List<String> nodeConfig, Tree tree) {
        int treeID = nodeConfig.size();
        nodeConfig.add("");

        StringBuilder stringBuilder = new StringBuilder();
        if (tree.children.isEmpty()) {
            String value = ((Terminal) tree.node).getParsedValue();
            stringBuilder
                    .append("label=\"")
                    .append(value.isEmpty() ? "[EPS]" : value)
                    .append("\" shape=box color=green");
            nodeConfig.set(treeID, stringBuilder.toString());
            return;
        }

        stringBuilder
                .append("label=\"")
                .append(((NonTerminal) tree.node).getValue())
                .append("\" shape=circle");
        nodeConfig.set(treeID, stringBuilder.toString());

        for (Tree child : tree.children) {
            int childID = nodeConfig.size();
            edges.add(new Edge(treeID, childID));
            walkthroughGraphvizImpl(edges, nodeConfig, child);
        }
    }
}
